character api added

iamananya opened the pull request:
- /character/list: API to get a list of user-owned characters
Added this API for Step 6 
Please review this @GuiltyMorishita

GuiltyMorishita commented on pkg/controllers/gacha-controller.go:
```diff
@@ -120,3 +121,44 @@ func UpdateUser(w http.ResponseWriter, r *http.Request) {
 
 }
+
+func CreateCharacter(w http.ResponseWriter, r *http.Request) {
```
CreateCharacter API is not defined in swagger.yaml.
What do you need it for?

iamananya replied:
To add the characters, we can also add it directly through the mysql `insert` statement as well. I created it to make request directly from postman and view it.

GuiltyMorishita replied:
@iamananya 
Imagine a program that you would actually use in production.
It's not good for users to be able to add characters on their own.
Please insert the character data as seed data at the same time as migrating the database.

iamananya replied:
Okay I will remove this api function

GuiltyMorishita commented on pkg/routes/gacha-routes.go:
```diff
@@ -10,6 +10,10 @@ var RegisterUserRoute = func(router *mux.Router) {
 	router.HandleFunc("/user/{userId}", controllers.GetUserById).Methods("GET")
 	router.HandleFunc("/user/{userId}", controllers.UpdateUser).Methods("PUT")
+	router.HandleFunc("/characters/", controllers.CreateCharacter).Methods("POST")
+	router.HandleFunc("/characters/list/", controllers.GetCharacters).Methods("GET")
```
The name of the controller should be ListCharacters. Get is used for singular.
ref: https://cloud.google.com/apis/design/standard_methods

iamananya replied:
Okay I will update accordingly.

GuiltyMorishita commented:
@iamananya 
There seems to be no user authentication function in the API in general.
Authenticate the user using the x-token header as per swagger.yaml.

Please make a PR for this support in another branch.

iamananya commented:
Okay I will make another PR supporting this.

iamananya commented:
I have added the api for `/gacha/draw/` and the changes have been made as suggested previously
* The name of the controller is changed to ListCharacters .
* CreateCharacter API has been removed.
Please review the changes @GuiltyMorishita .

GuiltyMorishita commented on pkg/models/gacha.go:
```diff
@@ -12,8 +12,50 @@ var db *gorm.DB
+}
+
+type GachaDrawRequest struct {
+	NumTrials int `json:"num_trials"`
```
According to the definition in swagger.yaml, this should be times, not num_trials.

iamananya replied:
Updated in the latest commit.

GuiltyMorishita commented on pkg/models/gacha.go:
```diff
@@ -56,8 +101,93 @@ func GetUserById(Id int64) (*User, *gorm.DB) {
+}
+
+func DrawCharacter(characters []Character) Character {
+	rand.Seed(time.Now().UnixNano())
```
math/rand is not appropriate, you can use crypto/rand.

iamananya replied:
Updated

iamananya replied:
Updated in the latest commit.

GuiltyMorishita commented on pkg/models/gacha.go:
```diff
@@ -56,8 +101,93 @@ func GetUserById(Id int64) (*User, *gorm.DB) {
+func DrawCharacter(characters []Character) Character {
+	rand.Seed(time.Now().UnixNano())
+
+	var rarityPool []Character
```
Isn't it a characterPool because it is an array of characters?

iamananya replied:
Updated.

iamananya replied:
Updated in the latest commit.

iamananya commented:
@GuiltyMorishita I have updated the code
I am generating the character pool once and then drawing characters from it which reduces the time complexity.
I have also updated the headers with the x-token parameter.
The parameters have been named accordingly and the crypto/rand library has been used in place of math/rand.

GuiltyMorishita commented:
@iamananya 

> There seems to be no user authentication function in the API in general.
Authenticate the user using the x-token header as per swagger.yaml.
Please make a PR for this support in another branch.

I asked for a separate PR to implement the user authentication
Make sure you have a reviewer friendly PR!

GuiltyMorishita commented on pkg/controllers/gacha-controller.go:
```diff
@@ -100,23 +85,112 @@ func CreateUser(w http.ResponseWriter, r *http.Request) {
+	w.Write(respBody)
+
+}
+func generatecharacterPool(characters []models.Character) []models.Character {
```
C should be capitalized.

iamananya replied:
@GuiltyMorishita I have updated it in the latest commit.

GuiltyMorishita commented on pkg/models/gacha.go:
```diff
@@ -22,42 +69,87 @@ func init() {
-	return Users
+func GetUserByToken(token string) []User {
+	var users []User
+	db.Where("token = ?", token).Find(&users)
```
Was the token unique?
I think multiple users would be returned if there were duplicates.

iamananya replied:
yes the token which I generated later was unique.

GuiltyMorishita replied:
@iamananya Where is it guaranteed to be unique?

iamananya replied:
@GuiltyMorishita  I have used a `generateRandomString` function so, its very unlikely that a token is ever repeated.

GuiltyMorishita commented on pkg/models/gacha.go:
```diff
@@ -22,42 +69,87 @@ func init() {
+	return users
+}
+
+func GetAllUsers(token string) []User {
```
Is there situation where multiple users are returned with a single token?

iamananya replied:
@GuiltyMorishita  Yes initially I had just implemented with random characters  from user input in that case multiple users are returned for same token. However I later generated a token of 20 characters string which Is not  repeating in that case it gives unique output.

GuiltyMorishita replied:
@iamananya If it is clear that it is a unique Token, there is no need for a method to retrieve multiple users using a token.

iamananya replied:
Okay.

iamananya replied:
@GuiltyMorishita  I have commented that portion of code in  the latest commit for retrieving multiple users using a single token.

GuiltyMorishita commented:
@iamananya Sorry for the rework, since this is a practice, can you make separate PR for authentication?

iamananya commented:
I am sorry that I didn't follow the instruction @GuiltyMorishita  :cry: .Yes I can make a separate PR for authentication.

GuiltyMorishita commented:
LGTM
